packages/animations/lib/src/dual_transition_builder.dart

```diff
@@ -0,0 +1,228 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+/// Builder callback used by [DualTransitionBuilder].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+/// The builder will be called whenever the [Animation.value] of `animation`
+/// has changed.
+typedef TransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// A animated builder that animates its [child] with different transitions
+/// based on the [AnimationStatus] of the provided [animation].
+///
+/// While the [animation] runs forward, the [child] is animated according to
+/// [forwardBuilder] and while the [animation] is running in reverse, it is
+/// animated according to [reverseBuilder].
+///
+/// This widget can be used to specify different enter and exit transitions for
+/// a [child].
+class DualTransitionBuilder extends StatefulWidget {
+  /// Creates a [DualTransitionBuilder].
+  ///
+  /// The [animation], [forwardBuilder], and [reverseBuilder] arguments are
+  /// required and must not be null.
+  const DualTransitionBuilder({
+    Key key,
+    @required this.animation,
+    @required this.forwardBuilder,
+    @required this.reverseBuilder,
+    this.child,
+  })  : assert(animation != null),
+        assert(forwardBuilder != null),
+        assert(reverseBuilder != null),
+        super(key: key);
+
+  /// The animation that drives the [child]'s transition.
+  ///
+  /// When this animation runs forward, the [child] transitions as specified by
+  /// [forwardBuilder]. When it runs in reverse, the child transitions according
+  /// to [reverseBuilder].
+  final Animation<double> animation;
+
+  /// A builder for a transition that wraps [child].
```

Comment:
Should we make the first sentence here and for `reverseBuilder` more specific? It might be confusing to readers to have to get to the 3rd paragraph to understand how this will be used:
```
/// A builder to create the transition that wraps [child] and makes 
/// [child] appear on screen.
```
and
```
/// A builder to create the transition that wraps [child] and makes 
/// [child] disappear from the screen.
```